### PR TITLE
Refacto/eosu 1007 replace eos config tests with product config-->6.1.0

### PR DESCRIPTION
**Pull request by OswaldoApexSystems**
[EOSU 1007 Update Config/EOSConfigTests.cs](https://epicgames.atlassian.net/browse/EOSU-1007)
Screenshot 2026-03-27 002928

This change migrates the configuration system from the legacy EOSConfig
model to the new ProductConfig-based architecture.

Key points of this migration:

- Replaced EOSConfig usage with ProductConfig as the main configuration
  entry point across runtime and tests.
- Refactored validation logic to rely on strong typing and domain
  invariants instead of string parsing.
- Introduced dedicated FieldValidators aligned with the new model:
  * GUIDFieldValidatorAttribute for Guid-based fields (rejects Guid.Empty).
  * SandboxIdFieldValidatorAttribute for SandboxId value object validation.
  * ProductionEnvironmentsFieldValidatorAttribute to validate deployment
    aggregates.
  * ClientCredentialsFieldValidatorAttribute to ensure at least one
    complete client credential exists.
- Removed legacy string-based ID validation patterns in favor of
  semantic validation using Guid, SandboxId, and aggregate state.

Domain model improvements:
- Deployment validity is now expressed exclusively through domain
  invariants (Deployment.IsComplete) instead of FieldValidator usage.
- SandboxId was hardened to be null-safe and robust against invalid input,
  preventing runtime exceptions and clarifying the distinction between
  "valid", "invalid", and "empty" states.
- Validation responsibilities are now clearly split:
  * FieldValidators validate configuration completeness at the
    ProductConfig boundary.
  * Domain objects (Deployment, SandboxId, Credentials) enforce their
    own invariants internally.

Test suite updates:
- Removed or migrated EOSConfig tests to ProductConfigTests.
- Added comprehensive test coverage for:
  * ProductConfig fields and validators.
  * ProductionEnvironments and ClientCredentials aggregates.
  * Deployment domain invariants.
  * SandboxId value object behavior, including edge cases, rollback
    behavior, and format semantics.
- Ensured tests are deterministic by explicitly resetting singleton
  config state between test cases.

Overall, this commit modernizes the configuration system by:
- Eliminating legacy parsing-based validation.
- Making invalid states unrepresentable through strong typing.
- Aligning runtime behavior, validation, and tests around the same
  domain rules.

### CODE REVIEW

File: com.playeveryware.eos/Runtime/Core/Config/Attributes/GUIDFieldValidatorAttribute.cs [hunk 1]
```diff
@@ -23,31 +23,22 @@
...
-            if (string.IsNullOrEmpty(stringValue))
+            if (!(toValidate is Guid guidValue))
             {
-                configurationProblemMessage = "The field value is an empty string.";
+                configurationProblemMessage = "The field value is not of type Guid.";
                 return false;
             }
 
-            if (!Guid.TryParse(stringValue, out Guid result))
+            if (Guid.Equals(guidValue,Guid.Empty))
```
Comment (matt-clarke) [inline com.playeveryware.eos/Runtime/Core/Config/Attributes/GUIDFieldValidatorAttribute.cs]:
```suggestion
            if (Guid.Equals(guidValue, Guid.Empty))
```

File: com.playeveryware.eos/Runtime/Core/Config/Attributes/SandboxIDFieldValidatorAttribute.cs [hunk 2]
```diff
@@ -23,39 +23,27 @@
...
             }
 
-            if (Regex.IsMatch(singleStringValue, PreProductionEnvironmentRegex))
-            {
-                configurationProblemMessage = string.Empty;
-                return true;
-            }
-
-            if (Guid.TryParse(singleStringValue, out _))
+            if (sandboxId.IsValid()||sandboxId.IsEmpty)
```
Comment (matt-clarke) [inline com.playeveryware.eos/Runtime/Core/Config/Attributes/SandboxIDFieldValidatorAttribute.cs]:
```suggestion
            if (sandboxId.IsValid() || sandboxId.IsEmpty)
```

File: com.playeveryware.eos/Runtime/Core/Config/SandboxId.cs [hunk 3]
```diff
@@ -86,6 +86,11 @@ readonly get
 
         public bool IsValid()
         {
+            if (string.IsNullOrEmpty(_value))
```
Comment (matt-clarke) [inline com.playeveryware.eos/Runtime/Core/Config/SandboxId.cs]:
```suggestion
            if (string.IsNullOrWhiteSpace(_value))
```

File: com.playeveryware.eos/Runtime/Core/Config/Attributes/SandboxIDFieldValidatorAttribute.cs [hunk 4]
```diff
@@ -23,39 +23,27 @@
...
             {
                 configurationProblemMessage = string.Empty;
                 return true;
             }
 
-            configurationProblemMessage = FieldDidNotMatchMessage;
+            configurationProblemMessage = InvalidSanboxIdMessage;
```
Comment (matt-clarke) [inline com.playeveryware.eos/Runtime/Core/Config/Attributes/SandboxIDFieldValidatorAttribute.cs]:
```suggestion
            configurationProblemMessage = InvalidSandboxIdMessage;
```

File: com.playeveryware.eos/Runtime/Core/Config/Attributes/SandboxIDFieldValidatorAttribute.cs [hunk 5]
```diff
@@ -23,39 +23,27 @@
...
-    using System.Text;
-    using System.Text.RegularExpressions;
 
     [AttributeUsage(AttributeTargets.Field)]
-
     public class SandboxIDFieldValidatorAttribute : FieldValidatorAttribute
     {
-        const string PreProductionEnvironmentRegex = @"^p\-[a-zA-Z\d]{30}$";
-        public const string FieldDidNotMatchMessage = "The field value is not a GUID, and did not match the regex used for Pre Production Environments: '" + PreProductionEnvironmentRegex + "'.";
+        public const string InvalidSanboxIdMessage = "SandboxId is invalid or empty.";
```
Comment (matt-clarke) [inline com.playeveryware.eos/Runtime/Core/Config/Attributes/SandboxIDFieldValidatorAttribute.cs]: Does this need to be public? Also fixing spelling mistake
```suggestion
        private const string InvalidSandboxIdMessage = "SandboxId is invalid or empty.";
```

File: com.playeveryware.eos/Runtime/Core/Config/Attributes/ClientCredentialsFieldValidatorAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        public const string NoValidClientCredentialsMessage =
+                "At least one complete Client Credential must be configured.";
+
+        public override bool FieldValueIsValid(object toValidate, out string configurationProblemMessage)
+        {
+            if (toValidate is not SetOfNamed<EOSClientCredentials> clients)
+            {
+                configurationProblemMessage =
+                    "Field value is not a set of EOSClientCredentials.";
```
Comment (matt-clarke) [inline com.playeveryware.eos/Runtime/Core/Config/Attributes/ClientCredentialsFieldValidatorAttribute.cs]:
```suggestion
                configurationProblemMessage = $"Field value type is not {nameof(SetOfNamed<EOSClientCredentials>)}.";
```

File: com.playeveryware.eos/Runtime/Core/Config/SandboxId.cs
Comment (matt-clarke) [inline com.playeveryware.eos/Runtime/Core/Config/SandboxId.cs]: Can you change this to IsNullOrWhiteSpace to detect the case that it's a string with just whitespace in

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+        /// </summary>
+        /// <typeparam name="T">The kind of validator attribute that is expected.</typeparam>
+        /// <param name="fieldName">The name of the field that is being checked for.</param>
+        /// <param name="failures">A list of all failures from validation.</param>
+        /// <param name="message">
+        /// A specific error message to check for.
+        /// Optional. If null or empty, this is not used.
+        /// </param>
+        /// <returns>True if the expected failure is within the list.</returns>
+        private bool failuresIncludeExpectedFailure<T>(string fieldName, List<FieldValidatorFailure> failures, string message = "") where T : FieldValidatorAttribute
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs]:
```suggestion
        private bool FailuresIncludeExpectedFailure<T>(string fieldName, List<FieldValidatorFailure> failures, string message = "") where T : FieldValidatorAttribute
```

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+    using PlayEveryWare.Common;
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using UnityEngine;
+
+    public class ProductConfigTests
+    {
+        [Test]
+        public void SandboxId_IsInvalid_WhenValueIsNull()
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs:35]: Could you move the SandboxId tests to a separate file? Although it's used by the ProductConfig it's not tied to it

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+        [Test]
+        public void SandboxId_IsInvalid_WhenValueIsEmptyString()
+        {
+            SandboxId sandboxId = new SandboxId
+            {
+                Value = string.Empty
+            };
+
+            Assert.IsFalse(sandboxId.IsValid(), "SandboxId should not be valid when its value is empty.");
+        }
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs:51]:
```suggestion
        }
        
        [Test]
        public void SandboxId_IsInvalid_WhenValueIsWhiteSpace()
        {
            SandboxId sandboxId = new SandboxId
            {
                Value = " "
            };

            Assert.IsFalse(sandboxId.IsValid(), "SandboxId should not be valid when its value is whitespace only.");
        } 
```

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+
+            Assert.DoesNotThrow(() =>
+            {
+                sandboxId.Value = "not-a-sandbox-id";
+            });
+        }
+
+
+        [Test]
+        public void ProductNameM_MustNotBeEmpty()
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs]:
```suggestion
        public void ProductName_MustNotBeEmpty()
```

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+            if (!FieldValidator.TryGetFailingValidatorAttributes(productConfig, out List<FieldValidatorFailure> failingAttributes))
+            {
+                Assert.Fail($"Config should have failing attributes.");
+            }
+
+            Assert.IsTrue(failuresIncludeExpectedFailure<NonEmptyStringFieldValidatorAttribute>(nameof(ProductConfig.ProductName),
+                failingAttributes,
+                NonEmptyStringFieldValidatorAttribute.FieldIsEmptyMessage),
+                "There should be a failure of the expected type and message.");
+
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs]:
```suggestion
```

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+        }
+
+        [Test]
+        public void ProductName_SuccessfulParsing()
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs]: Minor thing but these tests don't quite match the same format as the ones above
```suggestion
        public void ProductName_PassesValidation_WhenValueIsValid()
```

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+
+            Assert.IsFalse(failuresIncludeExpectedFailure<NonEmptyStringFieldValidatorAttribute>(
+                nameof(ProductConfig.ProductName),
+                failingAttributes,
+                NonEmptyStringFieldValidatorAttribute.FieldIsEmptyMessage),
+                "Product Name should not have errors describing it as an empty field.");
+        }
+
+        [Test]
+        public void ProductVersion_MustNotBeEmpty()
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs]: As above
```suggestion
        public void ProductVersion_FailsValidation_WhenValueIsEmpty()
```

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+
+            Assert.IsFalse(failuresIncludeExpectedFailure<NonEmptyStringFieldValidatorAttribute>(
+                nameof(ProductConfig.ProductVersion),
+                failingAttributes,
+                NonEmptyStringFieldValidatorAttribute.FieldIsEmptyMessage),
+                "Product Version should not have errors describing it as an empty field.");
+        }
+        /// <summary>
+        /// 
+        /// </summary>
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs]:
```suggestion

```

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+
+        [Test]
+        public void ClientCredentials_WithValidClient_ShouldPass()
+        {
+            ProductConfig productConfig = ProductConfig.Get<ProductConfig>();
+
+            productConfig.Clients.Add(new EOSClientCredentials(
+                "validClientId",
+                "validClientSecret",
+                null));
```
Comment (matt-clarke) [inline Assets/Tests/PlayMode/Config/ProductConfigTests.cs:349]: Seems a bit weird that the encryption key can be null, does that work fine?

**Review by matt-clarke — CHANGES_REQUESTED** [review]

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs (hunk 8 above)
Comment (OswaldoApexSystems) [reply Assets/Tests/PlayMode/Config/ProductConfigTests.cs:35]: I could create a separate file, but since this is such a basic part of product configuration, I think it should stay here without problems.

File: Assets/Tests/PlayMode/Config/ProductConfigTests.cs (hunk 15 above)
Comment (OswaldoApexSystems) [reply Assets/Tests/PlayMode/Config/ProductConfigTests.cs:349]: In modern versions of the plugin, an EncryptionKey is not required

**Review by matt-clarke — APPROVED** [review]

**Review by billfr-ApexSystems — APPROVED** [review]
